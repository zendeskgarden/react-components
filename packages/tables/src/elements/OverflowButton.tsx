/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { ButtonHTMLAttributes, useState } from 'react';
import PropTypes from 'prop-types';
import { composeEventHandlers } from '@zendeskgarden/container-utilities';
import OverflowStrokeIcon from '@zendeskgarden/svg-icons/src/16/overflow-stroke.svg';
import {
  StyledOverflowButton,
  IStyledOverflowButtonProps,
  StyledOverflowButtonIconWrapper
} from '../styled';
import { useTableContext } from '../utils/useTableContext';

/**
 * @extends ButtonHTMLAttributes<HTMLButtonElement>
 */
export const OverflowButton = React.forwardRef<
  HTMLButtonElement,
  Omit<IStyledOverflowButtonProps, 'size'> & ButtonHTMLAttributes<HTMLButtonElement>
>(({ onFocus, onBlur, isFocused: focused, ...other }, ref) => {
  const [isFocused, setIsFocused] = useState(false);
  const { size } = useTableContext();

  return (
    <StyledOverflowButton
      onFocus={composeEventHandlers(onFocus, () => {
        setIsFocused(true);
      })}
      onBlur={composeEventHandlers(onBlur, () => {
        setIsFocused(false);
      })}
      size={size}
      isFocused={typeof focused === 'undefined' ? isFocused : focused}
      ref={ref}
      {...other}
    >
      <StyledOverflowButtonIconWrapper>
        <OverflowStrokeIcon />
      </StyledOverflowButtonIconWrapper>
    </StyledOverflowButton>
  );
});

OverflowButton.displayName = 'OverflowButton';

OverflowButton.propTypes = {
  isHovered: PropTypes.bool,
  isActive: PropTypes.bool,
  isFocused: PropTypes.bool
};
