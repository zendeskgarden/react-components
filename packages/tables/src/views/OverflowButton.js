/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import TableStyles from '@zendeskgarden/css-tables';
import { retrieveComponentStyles } from '@zendeskgarden/react-theming';
import { composeEventHandlers } from '@zendeskgarden/container-utilities';

const COMPONENT_ID = 'tables.overflow_button';

export const StyledOverflowButton = styled.button.attrs(props => ({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  type: 'button',
  className: classNames(TableStyles['c-table__row__cell__overflow'], {
    [TableStyles['is-hovered']]: props.hovered,
    [TableStyles['is-active']]: props.active,
    [TableStyles['is-focused']]: props.focused
  })
}))`
  ${props => retrieveComponentStyles(COMPONENT_ID, props)};
`;

/**
 * Accepts all `<button>` props
 */
// eslint-disable-next-line react/prop-types
const OverflowButton = React.forwardRef(({ onFocus, onBlur, ...other }, ref) => {
  const [isFocused, setIsFocused] = useState(false);

  return (
    <StyledOverflowButton
      onFocus={composeEventHandlers(onFocus, () => {
        setIsFocused(true);
      })}
      onBlur={composeEventHandlers(onBlur, () => {
        setIsFocused(false);
      })}
      focused={isFocused}
      ref={ref}
      {...other}
    >
      &nbsp;
    </StyledOverflowButton>
  );
});

OverflowButton.propTypes = {
  hovered: PropTypes.bool,
  active: PropTypes.bool,
  focused: PropTypes.bool
};

export default OverflowButton;
