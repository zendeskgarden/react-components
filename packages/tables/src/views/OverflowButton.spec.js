/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { render, fireEvent } from 'garden-test-utils';

import OverflowButton from './OverflowButton';

describe('OverflowButton', () => {
  it('applies default styling by default', () => {
    const { container } = render(<OverflowButton />);

    expect(container.firstChild).toHaveClass('c-table__row__cell__overflow');
  });

  it('applies hovered styling if provided', () => {
    const { container } = render(<OverflowButton hovered />);

    expect(container.firstChild).toHaveClass('is-hovered');
  });

  it('applies active styling if provided', () => {
    const { container } = render(<OverflowButton active />);

    expect(container.firstChild).toHaveClass('is-active');
  });

  it('applies focused styling if provided', () => {
    const { container } = render(<OverflowButton focused />);

    expect(container.firstChild).toHaveClass('is-focused');
  });

  describe('onFocus', () => {
    it('applies focused state', () => {
      const { container } = render(<OverflowButton />);

      fireEvent.focus(container.firstChild);
      expect(container.firstChild).toHaveClass('is-focused');
    });
  });

  describe('onBlur', () => {
    it('removes focused state', () => {
      const { container } = render(<OverflowButton />);

      fireEvent.focus(container.firstChild);
      fireEvent.blur(container.firstChild);
      expect(container.firstChild).not.toHaveClass('is-focused');
    });
  });
});
