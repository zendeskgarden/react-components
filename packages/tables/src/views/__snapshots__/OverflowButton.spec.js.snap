// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverflowButton applies active styling if provided 1`] = `
<OverflowButton
  active={true}
>
  <OverflowButton__StyledOverflowButton
    active={true}
    focused={false}
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <button
      className="c-table__row__cell__overflow is-active "
      data-garden-id="tables.overflow_button"
      data-garden-version="version"
      onBlur={[Function]}
      onFocus={[Function]}
      type="button"
    >
       
    </button>
  </OverflowButton__StyledOverflowButton>
</OverflowButton>
`;

exports[`OverflowButton applies default styling by default 1`] = `
<OverflowButton>
  <OverflowButton__StyledOverflowButton
    focused={false}
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <button
      className="c-table__row__cell__overflow "
      data-garden-id="tables.overflow_button"
      data-garden-version="version"
      onBlur={[Function]}
      onFocus={[Function]}
      type="button"
    >
       
    </button>
  </OverflowButton__StyledOverflowButton>
</OverflowButton>
`;

exports[`OverflowButton applies focused styling if provided 1`] = `
<OverflowButton
  focused={true}
>
  <OverflowButton__StyledOverflowButton
    focused={true}
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <button
      className="c-table__row__cell__overflow is-focused "
      data-garden-id="tables.overflow_button"
      data-garden-version="version"
      onBlur={[Function]}
      onFocus={[Function]}
      type="button"
    >
       
    </button>
  </OverflowButton__StyledOverflowButton>
</OverflowButton>
`;

exports[`OverflowButton applies hovered styling if provided 1`] = `
<OverflowButton
  hovered={true}
>
  <OverflowButton__StyledOverflowButton
    focused={false}
    hovered={true}
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <button
      className="c-table__row__cell__overflow is-hovered "
      data-garden-id="tables.overflow_button"
      data-garden-version="version"
      onBlur={[Function]}
      onFocus={[Function]}
      type="button"
    >
       
    </button>
  </OverflowButton__StyledOverflowButton>
</OverflowButton>
`;
